Drop SYS_RESOURCE capability

ormergi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Currently virt-launcher runs with SYS_RESOURCE capability on VMs with SRIOV devices when SRIOVLiveMigation feature-gate enabled.
Having this capability enables Libvirt to adjust QEMU process memory lock limits (memlock rlimit) in case its already exists
As occurs when SRIOV VM is migrated and the SRIOV host-device is re-attached to VM on target.

With Libvirt  and above, Libvirt does best effort to adjusts QEMU process limits and wont try to change the limits if they are high enough.
In order to prevent using SYS_RESOURCE cap on containerized environments [1] as Kubevirt use case.

With this PR changes Kubevirt will adjust QEMU process rlimits as it does for libvirtd.
This will prevent libvirt from adjusting QEMU process rlimits (it will detect that limits are sufficient).
Then virt-launcher resources will be adjusted on post migration, in order to enable hot plugging VFIO devices

This PR also remove SYS_RESOURCE capability.
    
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Rebased on #5328 

`virt-handler `is privileged while` virt-launcher` is not.
Before this change, it was already done for the libvirtd process, the only difference here is that we now need the qemu one.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Drop virt-launcher SYS_RESOURCE capability
```

ormergi commented:
/test pull-kubevirt-e2e-kind-1.17-sriov

ormergi commented:
/test pull-kubevirt-e2e-kind-1.17-sriov

ormergi commented:
/test pull-kubevirt-e2e-kind-1.17-sriov

ormergi commented:
/cc @EdDev

EdDev commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -134,15 +134,20 @@ func (s *socketBasedIsolationDetector) Detect(vm *v1.VirtualMachineInstance) (Is
 		uintptr(pid),
-		limit,
-		uintptr(unsafe.Pointer(rlimit)), // #nosec used in unix RawSyscall6
+		uintptr(unix.RLIMIT_MEMLOCK), // #nosec used in unix RawSyscall6
```
Why the comment of `#nosec` moved to a different command? Are you sure it is not needed now for both now?

ormergi replied:
`#nosec` comments are added where unsafe pointers are used https://github.com/kubevirt/kubevirt/pull/4446
This comment no longer needed, I will remove it.

ormergi replied:
I just realized we still need this comment for line 145, I will just move it

EdDev commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -98,13 +99,19 @@ func (s *socketBasedIsolationDetector) DetectForSocket(vm *v1.VirtualMachineInst
 
+	if process, err := ps.FindProcess(pid); err != nil {
+		return nil, err
+	} else {
```
It is common in the code not to do this. How about:
```
process, err := ps.FindProcess(pid)
if err != nil {
    return ...
}
ppid := process.PPid()
```

EdDev reviewed (changes requested):
> gracefully shutdown in case of virt-launcher crash [1][2].

The links are not permanent, so these will become incorrect when the code change.

EdDev reviewed (commented):
< With Libvirt 7 (Coper)

I do not understand how the `Coper` part relevant here? That is the source of the package, we mainly care about the version. If you want to be specific, give a more detailed version.

EdDev reviewed (changes requested):
commit `post migration: Adjust QEMU and libvirt processes memlock limits`:

The change in this commit includes both logic addition and refactoring of the existing code. Please split the refactoring (and naming changes) to a separate commit so it will be easier to see what logic changed.

EdDev commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -164,43 +168,81 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
-		if err != nil {
-			return err
+func filterVirtLaunchersLibvirtProcesses(isolationResult IsolationResult, processes []ps.Process) []ps.Process {
+	var filteredProcesses []ps.Process
```
Why do you need this? On line 208 you can just use `:=`, no?

ormergi replied:
It may look redundant here because there are only two assignments.
I think its easier to track where a variable was declared when there are more then one assignments (the first is always `:=`).

EdDev commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -164,43 +168,81 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+func filterVirtLaunchersLibvirtProcesses(isolationResult IsolationResult, processes []ps.Process) []ps.Process {
+	var filteredProcesses []ps.Process
+
+	// consider all processes that are virt-launcher children
```
Can you please explain it all as a doctext? Having inline comments is hard to keep up-to-date and read the code.

EdDev commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -164,43 +168,81 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+
+	filteredProcesses = filterProcessesByParentPID(processes, []int{launcherPid, launcherParentPid})
+	// libvirtd process sets the memory lock limit before fork/exec-ing into qemu
+	// Its necessary to change QMEU process limits if its already exists (e.g: post migration just before hostdev)
```
What if we are in a reconcile? Are you sure we can reach this only on post-migration?

ormergi replied:
We will reach this also from reconcile.
Do we want to restrict changing QEMU process limits only on post migration?
If so I can create `AdjsutResourcesOnPostMigration` function 
that will be called on `finalizeMigration` (instead of regular `AdjustResources`)
@EdDev What you think?

EdDev replied:
I think it is less risky if the change is done per the specific need known to us. The qemu process memlock limit needs to be changed only when the VMI has a VFIO device and the state is post-migration.
If this is open to other scenarios, we may get into corner cases where we should not be.

> If so I can create AdjsutResourcesOnPostMigration function

As I see it, a function name should express what it does, not who or when it is called.

ormergi replied:
I pushed new changes so that only QEMU process will be adjusted on post migration

oshoval commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -152,48 +168,91 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
-		}
+	// bump libvirtd and qemu processes locked memory limits
+	libvirtProcesses := filterVirtLaunchersLibvirtProcesses(isolationResult, processes)
+	log.Log.Info("debug: adjust memlock: filtered processes:")
```
there are some debug logs just fyi (i guess its known)

vladikr replied:
yup, please use logs verbosity, log.Log.V(3).Info

ormergi replied:
I added these logs just for debugging I was going to remove them

vladikr replied:
maybe it is still valuable to keep it? up to you, but I think if it was valuable for you to debug, it might be useful when things go wrong ;)

vladikr commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -152,48 +168,91 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
-			Cur: uint64(memlockSize),
-		}
-		err = prLimit(process.Pid(), unix.RLIMIT_MEMLOCK, &rLimit)
+	for _, process := range libvirtProcesses {
```
can you please combine this loop with the one on line 190?

vladikr commented on pkg/virt-handler/isolation/isolation.go:
```diff
@@ -152,48 +168,91 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+	libvirtProcesses := filterVirtLaunchersLibvirtProcesses(isolationResult, processes)
+	log.Log.Info("debug: adjust memlock: filtered processes:")
+	for _, process := range libvirtProcesses {
+		log.Log.Infof("debug: pid %d, ppid %d, exec %s", process.Pid(), process.PPid(), process.Executable())
```
please make the debug prints more informative. Something like: "found libvirt process with pid %d...."

vladikr commented on pkg/virt-handler/vm.go:
```diff
@@ -2641,6 +2641,13 @@ func (d *VirtualMachineController) finalizeMigration(vmi *v1.VirtualMachineInsta
 		return err
 	}
+
+	err = d.podIsolationDetector.AdjustResources(vmi)
```
In reality, a failure here doesn't affect the result of migration.
The migration will still show as succeeded. However, perhaps we do need to react differently in this case? 
Maybe, we should at least send an alert in case we didn't manage to plug the devices back.

ormergi replied:
You right, its wrong to return on `AdjustResources` error, thanks for the heads up.
I will log the error instead of returning, then in case a device is failed to plug it will be logged.

maiqueb reviewed (commented):
There's one thing I might be missing about the commit "isolation: Add PPID to isolation result"; won't the PPID of the forked `virt-launcher` process always be 1 ? I.e. it is the entrypoint of the container.

ormergi commented:
Rebase

ormergi commented:
> There's one thing I might be missing about the commit "isolation: Add PPID to isolation result"; won't the PPID of the forked `virt-launcher` process always be 1 ? I.e. it is the entrypoint of the container.

Correct, there is a mistake in the commit message the forked virt-launcher PID is wrong.
Sorry about that, I will fix it.

ormergi commented:
> There's one thing I might be missing about the commit "isolation: Add PPID to isolation result"; won't the PPID of the forked `virt-launcher` process always be 1 ? I.e. it is the entrypoint of the container.

@maiqueb Correct, there is a mistake in the commit message the forked virt-launcher PID is wrong.
Sorry about that, I will fix it.

ormergi commented:
> There's one thing I might be missing about the commit "isolation: Add PPID to isolation result"; won't the PPID of the forked `virt-launcher` process always be 1 ? I.e. it is the entrypoint of the container.

@maiqueb Yes this is correct (note that from the node perspective the virt-launcher PID is different)
There is a mistake in the commit message, the PID of libvirtd at the  processes state example is wrong 
I will fix it.

ormergi commented:
Changes:
- Address code review comments
- I reverted the changes to isolation `AdjustResources` function, and created new function for QEMU process adjustments
  Now only QEMU process memory limits will be adjusted on post migration
- Moved process filtering code to new file
-  Added unit tests

ormergi commented:
Rebase

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -161,6 +161,61 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
 	return nil
 }
 
+// SetQemuProcessMemoryLimits adjusts QMEU process memory locked rlimits
```
I will move this to line 167

EdDev commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -143,11 +143,9 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
-			Cur: uint64(memlockSize),
-		}
-		err = prLimit(process.Pid(), unix.RLIMIT_MEMLOCK, &rLimit)
+		memlockSizeUint64 := uint64(memlockSize)
+
```
nit: It is not nice to place the name of the type in the name and there is no need for the newline.
You can use here `s` or `size` as it is one line away.
You can also drop this, because you can do the same in `setProcessMemoryLockRLimit`.
But frankly, why it is not returned with the correct type from `getMemlockSize`?

EdDev commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -157,6 +155,25 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+		uintptr(unsafe.Pointer(&rlimit)), // #nosec used in unix RawSyscall6
+		0, 0, 0)
+	if errno != 0 {
+		return fmt.Errorf("failed to set process %d memory locked rlimit %+v: %v", pid, rlimit, errno)
```
nit: `memory locked` ==>> `memlock`.

EdDev commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -143,11 +143,9 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+
+		err = setProcessMemoryLockRLimit(process.Pid(), memlockSizeUint64, memlockSizeUint64)
 		if err != nil {
 			return fmt.Errorf("failed to set rlimit for memory lock: %v", err)
```
This is not repeating itself as the returned error includes this info.

EdDev commented on pkg/virt-handler/isolation/process.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// filterProcessByPPID returns processes that runs the given executable.
+func filterProcessByExecutable(processes []ps.Process, exectutable string) []ps.Process {
```
Looking at the usage, it is enough for this function to return the first detected process with the mentioned executable.
Generalizing it now just complicates things unnecessarily.

And in such a case, this can be named: `lookupProcessByExecutable`.

EdDev commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,69 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+	if err := setProcessMemoryLimitsFn(process.Pid(), memlockSize, memlockSize); err != nil {
+		return fmt.Errorf("failed to set rlimit for memory lock: %v", err)
+	}
+	log.Log.V(3).Object(vmi).Infof("changed process memory limits changed Process: %+v Cur: %d Max:%d",
```
Too many "changed" appearances.

ormergi replied:
Fixed

EdDev commented on pkg/virt-handler/isolation/detector_test.go:
```diff
@@ -27,7 +27,9 @@ import (
 	"path/filepath"
 
 	gomock "github.com/golang/mock/gomock"
+	"github.com/mitchellh/go-ps"
```
This does not fit in this import block.

ormergi replied:
Fixed

EdDev replied:
This seems to have been reverted back.

ormergi replied:
Done

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -2644,8 +2644,12 @@ func (d *VirtualMachineController) finalizeMigration(vmi *v1.VirtualMachineInsta
-	if err != nil {
+
+	if err := isolation.AdjustQemuProcessMemoryLimits(d.podIsolationDetector, vmi); err != nil {
+		log.Log.Object(vmi).Reason(err).Warning("failed to adjust QEMU process memory limits")
```
I think that the text should be `"failed to finalize migration"` and the `err` to contain the details (which I am pretty sure already exists).

ormergi replied:
Done,
Also changed the error messages of the related functions, no it will return proper error message with no duplication.

EdDev reviewed (changes requested):
Looks nice, thanks!

Note that I have not passed through the unit tests too much, I plan to do so after the main points in the production code have been resolved.

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,69 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+		return fmt.Errorf("failed to get all processes: %v", err)
+	}
+
+	qemuProcess := findIsolatedQemuProcess(processes, []int{isolationResult.Pid(), isolationResult.PPid()})
```
On virt-launcher we use isolation package in order to detect the correct virt-launcher pod details (e.g: pid)
according to a given VM.
The implementation that currently being used is `socketBasedIsolationDetector` https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-handler/virt-handler.go#L285

Both virt-launcher pid and ppid are passed to `findIsolatedQemuProcess` in order to detect QEMU process pid whether virt-launcher process is forked or not (runs with flag `no-fork`, `false` or `true` respectively),
due to how `socketBasedIsolationDetector` works:
- virt-launcher is not forked (`--no-fork == true`)
   IsolationDetector returns virt-launcher process pid
   QEMU process is child of virt-launcher.
   ```
   pid: 1 ppid: 0 cmd: virt-launcher
   |_ ...
   |_ pid 100 ppid: 1 cmd: libvirtd  <-------------- isolation detector scope
   |
   |_ pid: 200 ppid: 1 cmd: qemu-kvm
    ```
- virt-launcher process is forked (`--no-fork == false`)
  There will be two nested `virt-launcher` processes. where qemu process is child of the 
   top level virt-launcher process.
   IsolationDetector returns the **nested** `virt-launcher` process **pid**, but QEMU process is not a child of this 
   process
   Thus we need the "nested-virt-launcher" process ppid, in order to get the "top-level-virt-launcher" process child's
   and find QEMU process.
   ```
   pid: 1 ppid: 0 cmd: virt-launcher
   |
   |_ pid: 100 ppid: 1 cmd: virt-launcher 
   |   |
   |   |_pid 1100 ppid: 100 cmd: libvirtd  <-------------- isolation detector scope
   |
   |_ pid: 200 ppid: 1 cmd: qemu-kvm
    ```

EdDev replied:
As I understand it, there is only one way virt-launcher is executed, not two ways.
It always comes up as pid 1 in the container, with the `--no-fork` missing (i.e. it should fork itself). When it forkes itself, it uses this flag, therefore the recursion will not continue on endlessly.
So the first scenario cannot happen.

ormergi replied:
Make sense, I will change it to pass only virt-launcher ppid it would also answer to the comment about `filterProcessByPPid` implementation.

ormergi commented on pkg/virt-handler/isolation/process.go:
```diff
@@ -0,0 +1,51 @@
+
+// filterProcessByPPID returns chiled processes of the given PID's.
+func filterProcessByPPID(processes []ps.Process, pids []int) []ps.Process {
+	filterProcessParentPid := map[int]struct{}{}
```
Done

ormergi commented on pkg/virt-handler/isolation/process.go:
```diff
@@ -0,0 +1,51 @@
+	return filteredProcesses
+}
+
+// filterProcessByPPID returns processes that runs the given executable.
```
Fixed

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,69 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+const qemuProcessExecutable = "qemu-kvm"
+
+// AdjustQemuProcessMemoryLimits adjusts QMEU process MEMLOCK rlimits that runs inside
+// virt-launcher pod on the given VM according to VM spec.
```
Fixed

EdDev commented on pkg/virt-handler/isolation/detector_test.go:
```diff
@@ -143,6 +143,23 @@ var _ = Describe("Isolation Detector", func() {
 	})
+
+	Context("VM with VFIO device", func() {
+		It("should adjust QEMU process MEMLOCK rlimits", func() {
```
I think that except `findIsolatedQemuProcess`, the other functions are hard to test.

I do not have a good solution at this point, but we should probably look in a following refactor on how to resolve this nicer.
If other reviewers have a better idea, maybe we can do something else.

ormergi replied:
I just realized how wrong this is.. sorry.
I removed this awkward test with testing `findIsolatedQemuProcess` which is what make sense to test
And considering the current design is what possible to test.

maiqueb reviewed (commented):
A couple of suggestions, but overall, I think the PR looks good; the most concerning section was already identified by @EdDev - the stub to interface conversion introduced in the unit tests.

Plus, I also think that you should update the commit msg to your second commit - "isolation: Add PPID to isolation result". While the example you provide is helpful, I was mislead by it; I think you should grab the example process table from host / virt-handler PID namespace, not from the launcher. Just the relevant section, not the full table (that would be confusing).

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,159 @@
+	return result
+}
+
+func convertToPsGoProcess(process *ProcessStub) ps.Process {
```
For some reason the test panics with the following error when I dont use `convertToPsGoProcess`
I think its related to how ginkgo marchers works
```
      reflect: Call using isolation.ProcessStub as type ps.Process
```

EdDev replied:
You tried to pass `&procStub1`or `procStub1` directly and this is what you got?
Based on how you converted it now, you need the first.

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,159 @@
+				"processA",
+			),
+			table.Entry("valid set, zero elements to filter",
+				convertToPsGoProcessList([]ProcessStub{
```
Is it acceptable to use this function under Context/Describe block?

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,159 @@
+			),
+			table.Entry("valid set, zero elements to filter",
+				convertToPsGoProcessList([]ProcessStub{
+					{ppid: 1, pid: 120, binary: "processA"},
+					{ppid: 1, pid: 110, binary: "processC"},
+					{ppid: 110, pid: 2222, binary: "processB"},
+					{ppid: 110, pid: 3333, binary: "processD"},
```
FYI: Its not possible to define const array or struct

maiqueb replied:
Riiiiight! Sorry for that. 

Would a var or a function returning that make sense to you ? 

```golang

getEmptyProcessList := func() []ProcessStub {}
getProcessList := func() []ProcessStub {
        {ppid: 1, pid: 120, binary: "processA"},
        {ppid: 1, pid: 110, binary: "processC"},
	{ppid: 110, pid: 2222, binary: "processB"},
	{ppid: 110, pid: 3333, binary: "processD"},
}
```

ormergi replied:
Sure, thats what I was going to do :)

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,159 @@
+package isolation
+
+import (
+	"github.com/mitchellh/go-ps"
+
```
We place third-parties imports in their own section, isn't it?

maiqueb replied:
All I see are 4 github.com imports. 

Aren't the 3 below 3rd party ?

ormergi replied:
Yes but they related to test infra, I didnt explained myself correctly,
As far as I know there should be four sections of imports on tests:
- Go builtin
- Third-parties
- Test infra (e.g ginkgo)
- Kubevirt project

Please correct me if I am wrong here

maiqueb replied:
I honestly don't know. I think I've been using the test infra stuff mixed with the 3rd parties.

Have it your way; it's mostly the same.

ormergi replied:
Cool

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -143,20 +143,36 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
 		if err != nil {
 			return err
 		}
-		rLimit := unix.Rlimit{
```
Before this commit changes, the function `prLimit` represented  and worked like the binary tool `prLimit`
where you need to pass pid, resource type you want to limit (e.g memory, file-descriptors..) and size (rlimit struct).
We use this function only for adjusting memory limits and there is no benefit with before this commit changes granularity.
This commit make this function nicer by encapsulating the resource type and rlimit struct definition
it gives the reader better sense of what it does then before.

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -143,20 +143,36 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
-		}
-		err = prLimit(process.Pid(), unix.RLIMIT_MEMLOCK, &rLimit)
+
+		err = setProcessMemoryLockRLimit(process.Pid(), memlockSize, memlockSize)
```
Make sense, at the current logic we keep current and max limits the same
I will change it to have single size parameter

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,62 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+	if err := setProcessMemoryLimitsFn(process.Pid(), memlockSize, memlockSize); err != nil {
+		return fmt.Errorf("failed to set process %d memlock rlimit to %d: %v", process.Pid(), memlockSize, err)
+	}
+	log.Log.V(3).Object(vmi).Infof("set process %+v memlock rlimits to: Cur: %d Max:%d",
```
Gotcha, I will increase it to 5

ormergi commented on pkg/virt-handler/isolation/detector_test.go:
```diff
@@ -143,6 +143,23 @@ var _ = Describe("Isolation Detector", func() {
+			vmi := newMinimalVMIWithSingleSRIOVIface("testvmi", "sriov1")
+			virtLauncherPPid := 1
+			virtLauncherProcesses := convertToPsGoProcessList([]ProcessStub{
+				{pid: 1, ppid: 0, binary: "virt-launcher --no-fork false"},
```
Yes `--no-fork` default is false, I will remove it.

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,159 @@
+			func(processes []ps.Process, ppid int, expectedProcesses []ps.Process) {
+				Expect(childProcesses(processes, ppid)).To(ConsistOf(expectedProcesses))
+			},
+			table.Entry("empty set, zero elements to filter",
```
Got it, I will change the descriptions as you suggests.

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,159 @@
+			table.Entry("empty set, zero elements to filter",
+				[]ps.Process{},
+				10,
+				// expected
```
No problem, I will remove these and use variables instead.

ormergi commented:
Addressed code reviews comments:
- Changed `setProcessMemoryLockRLimit` (first commit) to get ps.Process and size instead of pid, current size and max size.
- Fixed typos and rephrased functions doctext and tests descriptions.
- Replaced the akward test I added to detector_test.go with test for `findIsolatedQemuProcess` 
which is what possible and make sense to test.

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,62 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+}
+
+// findIsolatedQemuProcess returns the chiled process of the given PID's that runs the QEMU process.
+// This function expect single process that runs QEMU, if there are more then one occurrences,
```
Done

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,62 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+	return adjustIsolatedProcessMemoryLimits(vmi, qemuProcess, setProcessMemoryLockRLimit)
+}
+
+// findIsolatedQemuProcess returns the chiled process of the given PID's that runs the QEMU process.
```
Yep, fixed :)

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,62 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+
+// adjustIsolatedProcessMemoryLimits adjusts the given process MEMLOCK rlimits
+// according to the given VMI spec
+func adjustIsolatedProcessMemoryLimits(vmi *v1.VirtualMachineInstance,
```
Thanks for the heads up, I removed this function.

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,62 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+	processes = childProcesses(processes, pid)
+	qemuProcess := lookupProcessByExecutable(processes, qemuProcessExecutable)
+
+	return qemuProcess
```
Done

ormergi commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,62 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+const qemuProcessExecutable = "qemu-kvm"
+
+// AdjustQemuProcessMemoryLimits adjusts QMEU process MEMLOCK rlimits that runs inside
+// virt-launcher pod on the given VMI according to VMI spec.
```
Done

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,159 @@
+				// expected
+				convertToPsGoProcessList([]ProcessStub{
+					{ppid: 110, pid: 2222, binary: "processB"},
+					{ppid: 110, pid: 3333, binary: "processC"},
```
Done

ormergi commented:
/retest

EdDev commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,134 @@
+var emptyProcessList = []ps.Process{}
+
+var _ = Describe("process", func() {
+	const proc1Andproc5Exec = "processA"
+	const proc2Andproc4Pid = 110
+	const nonExistPPid = 300
```
nit: We usually use a const block.
```
const (
...
)
```

I would also not embed two things into one name.
If we take `proc1Andproc5Exec`, then you can just call it: `processTestExecPath`.
Some for the pid one, processTestPID.

ormergi replied:
Done

EdDev commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,134 @@
+
+	Context("find child processes", func() {
+		proc2AndProc4StubList := []ProcessStub{procStub2, procStub4}
+		proc2AndProc4 := convertToPsGoProcessList(proc2AndProc4StubList)
```
I just wanted this to be done in the test function so we will have a simple function signature and do the conversions at the place it is needed, not before.

Now we have a global (`emptyProcessList`) and these two variables in the context container which are visible to all tests under it.
The simple path, is also the better choice here. Please move the conversion in the function and remove all these 3 redundant variables.

ormergi replied:
I removed `proc2AndProc4StubList` and `proc2AndProc4`, 
now their values are inline.
Moved `emptyProcessList` to the describe block,
And moved conversions calls to the test function.

Having `emptyProcessList` is more readable, isnt it?

EdDev commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,134 @@
+	})
+
+	Context("lookup process by executable", func() {
+		proc1 := convertToPsGoProcess(&procStub1)
+
+		procStub5 := ProcessStub{ppid: 100, pid: 220, binary: proc1Andproc5Exec}
+		procsListWithProcsRunningSameExec := convertToPsGoProcessList(append(procsStubList, procStub5))
```
same

ormergi replied:
I removed `proc1` and `procsListWithProcsRunningSameExec`, now their values are inline.
Also moved the conversion to the test function.

Having `procStub5` in its own variable is more readable instead of define it inline on the test function, isnt it?

EdDev commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,134 @@
+		procStub5 := ProcessStub{ppid: 100, pid: 220, binary: proc1Andproc5Exec}
+		procsListWithProcsRunningSameExec := convertToPsGoProcessList(append(procsStubList, procStub5))
+
+		table.DescribeTable("should return nil",
```
how about: "should find no process"

ormergi replied:
Done

EdDev commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,134 @@
+				Expect(lookupProcessByExecutable(processes, executable)).
+					To(Equal(expectedProcess))
+			},
+			table.Entry("given processes list that includes exactly one process that runs the given executable",
```
`runs the given executable` is repeating the text from the describe table. Please check it reads well when you run it with verbose.

ormergi replied:
I think removing the world given makes it better
```
process
  find child processes
    should return the first occurrence of a process that runs the given executable,
      given processes list that includes exactly one process that runs the executable
```

EdDev commented on pkg/virt-handler/isolation/detector_test.go:
```diff
@@ -155,3 +155,22 @@ var _ = Describe("getMemlockSize", func() {
+			virtLauncherForkedProc,
+			libvirtdProc,
+		})
+		qemuPsProcess := convertToPsGoProcess(&qemuProc)
```
You cannot pass `&qemuProc` instead of `qemuPsProcess` ?

ormergi replied:
Yep it works with `&qemuProc`, I removed `convertToPsGoProcess`.

EdDev commented on pkg/virt-handler/vm.go:
```diff
@@ -2640,13 +2640,18 @@ func (d *VirtualMachineController) updateDomainFunc(old, new interface{}) {
 }
 
 func (d *VirtualMachineController) finalizeMigration(vmi *v1.VirtualMachineInstance) error {
+	errorMessage := "failed to finalize migration"
```
nit: Can be a const.

ormergi replied:
Done

maiqueb commented:
Other than the commit msg of `isolation: Add PPID to isolation result` using the process table from the launcher & the unit tests process stub (which @EdDev is already asking you about), I'm OK with the state of this PR.

ormergi commented:
@maiqueb  I changed  the process table on commit `isolation: Add PPID to isolation result` message

phoracek commented:
/retest

maiqueb commented on pkg/virt-handler/vm.go:
```diff
@@ -2640,13 +2640,18 @@ func (d *VirtualMachineController) updateDomainFunc(old, new interface{}) {
-		log.Log.Object(vmi).Reason(err).Error("failed to finalize migration")
+
+	if err := isolation.AdjustQemuProcessMemoryLimits(d.podIsolationDetector, vmi); err != nil {
+		log.Log.Object(vmi).Reason(err).Warning(errorMessage)
```
Don't we want to return here ? (I honestly don't know).

ormergi replied:
Changing QEMU memlock rlimits doesn't affect the migration result, 
there was shot discussion about it  https://github.com/kubevirt/kubevirt/pull/5558#discussion_r625352942

vladikr replied:
Yep, we shouldn't fail the migration because of this.
However, @ormergi perhaps we should emit an event, so it will be visible on the vmi object and not buried in the logs.
More importantly [here](https://github.com/kubevirt/kubevirt/blob/8cfc7b32b62cd451396d47448083029b7313ad71/pkg/virt-launcher/virtwrap/live-migration-target.go#L38), I think

ormergi replied:
No problem I will add an event to `AdjustQemuProcessMemoryLimits` instead of waning log message
I prefer to add event to https://github.com/kubevirt/kubevirt/blob/8cfc7b32b62cd451396d47448083029b7313ad71/pkg/virt-launcher/virtwrap/live-migration-target.go#L38 
on a follow up PR if its ok with you

EdDev commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,120 @@
+
+	Context("lookup process by executable", func() {
+		procStub5 := ProcessStub{ppid: 100, pid: 220, binary: processTestExecPath}
+		//procsListWithProcsRunningSameExec := append(processes, procStub5)
```
leftover?

ormergi replied:
Yeah sorry I will remove it

EdDev commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,120 @@
+	procStub2 := ProcessStub{ppid: processTestPID, pid: 2222, binary: "processB"}
+	procStub3 := ProcessStub{ppid: 1, pid: 110, binary: "processC"}
+	procStub4 := ProcessStub{ppid: processTestPID, pid: 3333, binary: "processD"}
+	processes := []ProcessStub{procStub1, procStub3, procStub2, procStub4}
```
This name is used in the function argument as well. It is confusing as one shadows the other.

ormergi replied:
I will rename process -> testProcesses

maiqueb commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,120 @@
+
+		table.DescribeTable("should return the first occurrence of a process that runs the given executable",
+			func(processes []ps.Process, executable string, expectedProcess ps.Process) {
+				Expect(lookupProcessByExecutable((processes), executable)).
```
You have redundant parentheses around the `processes` variable.

ormergi replied:
I am going to move `convertToPsGoProcessList` calls to describe table function, I will also fix this.

maiqueb commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,119 @@
+	})
+})
+
+func convertToPsGoProcessList(processes []ProcessStub) []ps.Process {
+	result := []ps.Process{}
+	for idx := range processes {
+		result = append(result, &processes[idx])
+	}
+
+	return result
+}
```
You don't need this. You can directly use a `[]ps.Process{procStub1, ..., procStubN}` when defining the arrays.

ormergi replied:
Good catch! thanks for the heads up :)

maiqueb commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,119 @@
+	return p.pid
+}
+
+func (p *ProcessStub) PPid() int {
```
ditto

maiqueb commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,119 @@
+	return p.ppid
+}
+
+func (p *ProcessStub) Executable() string {
```
ditto

maiqueb commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,119 @@
+	procStub2 := ProcessStub{ppid: processTestPID, pid: 2222, binary: "processB"}
+	procStub3 := ProcessStub{ppid: 1, pid: 110, binary: "processC"}
+	procStub4 := ProcessStub{ppid: processTestPID, pid: 3333, binary: "processD"}
+	testProcesses := []ProcessStub{procStub1, procStub3, procStub2, procStub4}
```
Here you want a `[]ps.Process{}` , not a stub array: 

```golang
processes := []ps.Process{procStub1, procStub3, procStub2, procStub4}
```

ormergi replied:
Done

maiqueb commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,119 @@
+				testProcesses, nonExistPPid, emptyProcessList,
+			),
+			table.Entry("given process list and pid where there are child processes of the given pid",
+				testProcesses, processTestPID, []ProcessStub{procStub2, procStub4},
```
Here you want `[]ps.Process{procStub2, procStub4}` not this stub array.

ormergi replied:
Done

maiqueb reviewed (changes requested):
No need for the explicit type conversion.

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,119 @@
+	binary string
+}
+
+func (p *ProcessStub) Pid() int {
```
I followed how `ps-go` package implements other kinds of processes (e.g `UnixProcess`) where the receivers were pointers.
I can confirm that if the receivers are pointers then its not possible to define `ps.Process` array where the objects are from a different type (e.g `ProcessStub`).
Removing the pointers from the receivers makes it possible, I dont know to explain why though.

EdDev replied:
If the receiver is a pointer, you can populate the list with both pointers and values: https://play.golang.org/p/mVnwoCUmGT_q

If the receiver is a value, you can no longer use values in the list, you need pointers.

ormergi replied:
@maiqueb  I tested it and its like you said, Go compiler didn't recognized `ProcessStub` as implementation of `ps.Process`
because the receiver type of `ProcessStub` is pointer type. 
But on the tests `ProcessStub` objects are defined as non-pointer types.
I also removed the pointers from `ProcessStub` as we discussed there is no actual need for them to be pointers.
Thanks again for the heads up!

ormergi commented:
Changes:
- Removed conversion function from `process` and `detector` unit tests

ormergi commented on pkg/virt-handler/isolation/process_test.go:
```diff
@@ -0,0 +1,119 @@
+		processTestPID      = 110
+		nonExistPPid        = 300
+	)
+	emptyProcessList := []ProcessStub{}
```
Done

EdDev commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -160,6 +160,69 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
 
+const qemuProcessExecutable = "qemu-kvm"
+
+// AdjustQemuProcessMemoryLimits adjusts QMEU process MEMLOCK rlimits that runs inside
```
ping

ormergi replied:
Fixed

ormergi commented:
Changes:
- Fixed typos and import ordering on `detector.go` and `detector_test.go`

ormergi commented:
/test pull-kubevirt-generate pull-kubevirt-build-arm64

Both jobs failed due to timeout error when pulling kubevirt builder image

maiqueb reviewed (approved):
Code-wise, this looks good to me.

But I wonder what's the purpose of the `SRIOVMigration` feature gate now that you've removed the `SYS_RESOURCE` capability requirement from the launcher pod.

If we end up not wanting it, can be done on a follow up.

EdDev reviewed (approved):
Thank you

ormergi commented:
> But I wonder what's the purpose of the `SRIOVMigration` feature gate now that you've removed the `SYS_RESOURCE` capability requirement from the launcher pod.
> 
> If we end up not wanting it, can be done on a follow up.

Basically we want the feature-gate because SRIOV migration is not matured enough and we have more work to do.
For example our next task is to cover the case that if the migration is canceled then the SRIOV device should  re-attached to the original VM.

EdDev commented:
> > But I wonder what's the purpose of the `SRIOVMigration` feature gate now that you've removed the `SYS_RESOURCE` capability requirement from the launcher pod.
> > If we end up not wanting it, can be done on a follow up.
> 
> Basically we want the feature-gate because SRIOV migration is not matured enough and we have more work to do.
> For example our next task is to cover the case that if the migration is canceled then the SRIOV device should re-attached to the original VM.

The original reason is the same as the other features protected by the gates.
Usually, if a feature affects the overall system (e.g. migration), a feature-gate is placed to control it.
I am not sure if there is any written policy on this, but for this specific case it was an explicit request from the maintainers (I think it was @vladikr ).

maiqueb commented:
> > > But I wonder what's the purpose of the `SRIOVMigration` feature gate now that you've removed the `SYS_RESOURCE` capability requirement from the launcher pod.
> > > If we end up not wanting it, can be done on a follow up.
> > 
> > 
> > Basically we want the feature-gate because SRIOV migration is not matured enough and we have more work to do.
> > For example our next task is to cover the case that if the migration is canceled then the SRIOV device should re-attached to the original VM.
> 
> The original reason is the same as the other features protected by the gates.
> Usually, if a feature affects the overall system (e.g. migration), a feature-gate is placed to control it.
> I am not sure if there is any written policy on this, but for this specific case it was an explicit request from the maintainers (I think it was @vladikr ).

I had wrongly assumed the feature gate motivation originated from the extra capability requirement.

Thanks for the explanation.

AlonaKaplan commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -143,20 +143,38 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
 
+// setProcessMemoryLockRLimit Adjusts process MEMLOCK
+// soft-limit (current) and hard-limit (max) to the given size.
+func setProcessMemoryLockRLimit(process ps.Process, size int64) error {
```
Please pass just the `pid` and not the whole `process`.

ormergi replied:
Done

EdDev commented:
Please do not mix rebase with your changes, we have now no way to see the changes done without re-passing through all of them.

AlonaKaplan commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -159,6 +159,47 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
 	return nil
 }
 
+const qemuProcessExecutable = "qemu-kvm"
```
According to [this](https://github.com/kubevirt/kubevirt/blob/b2f3a6520ab821dbdf7a7a15f3da24bc32ed449a/cmd/virt-launcher/virt-launcher.go#L562) it can be `qemu-system` or `qemu-kvm`.

EdDev replied:
Maybe @vladikr can shade a light on this. We do not deploy `qemu-system` on u/s, but maybe some other d/s is?
Even if they do, we can never assure it really works as we do not test it.

vladikr replied:
yes, this can easily be `qemu-system`. our virt-launcher is based on an operating system that calls this executable qemu-kvm, but this can change.
Normally, we are looking for both, like in https://github.com/kubevirt/kubevirt/blob/master/cmd/virt-launcher/virt-launcher.go#L560
However, perhaps it's possible to look for a partial name `qemu` or make this configurable?

ormergi replied:
@vladikr  We could set an array with the names of qemu process than search for one of them, as follows:
```
var qemuProcessExecutables = []string{"qemu-system", "qemu-kvm"}

func findIsolatedQemuProcess(processes []ps.Process, pid int) (ps.Process, error) {
        processes = childProcesses(processes, pid)
       for _, exec := range qemuProcessExecutables {
               if qemuProcess := lookupProcessByExecutable(processes, exec); qemuProcess != nil 
                       return qemuProcess, nil
               }
       }

       return nil, fmt.Errorf("no QEMU process found under process %d child processes", pid)
}
```
I think its safer than looking for `qemu-*`, there could be other qmeu processes other than `qemu-kvm/system`?
What you think?

AlonaKaplan commented:
General question. Why libvirt needs `SYS_RESOURCE` to change the qemu rlimit but kubvirt can do it without this capability?

AlonaKaplan commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -159,6 +159,47 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+	if err != nil {
+		return fmt.Errorf("failed to get all processes: %v", err)
+	}
+	qemuProcess := findIsolatedQemuProcess(processes, isolationResult.PPid())
```
What if [no-fork](https://github.com/kubevirt/kubevirt/blob/b2f3a6520ab821dbdf7a7a15f3da24bc32ed449a/cmd/virt-launcher/virt-launcher.go#L345) is `true`. Would the virt-launcher process and the qemu one still have the same ppid?

EdDev replied:
That flag is a hack, we can reach here only in one condition. The binary is called without the  flag and then it calls itself with the flag (so not recursion will occur).

ormergi replied:
No, when no-fork is `true` there will be no extra virt-launcher process unlike when no-fork is `false`,
virt-launcher and qmeu processes will not have the same PPID.

There was a dissolution about this flag with more details about virt-launcher processes state
https://github.com/kubevirt/kubevirt/pull/5558#discussion_r629351421

AlonaKaplan commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -159,6 +159,47 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+	if err != nil {
+		return fmt.Errorf("failed to get all processes: %v", err)
+	}
+	qemuProcess := findIsolatedQemuProcess(processes, isolationResult.PPid())
```
Please return a relevant error if `qemuProcess` is `nil`.

ormergi replied:
Done

ormergi commented:
> General question. Why libvirt needs `SYS_RESOURCE` to change the qemu rlimit but kubvirt can do it without this capability?

On SRIOV VM's migration we detach the SRIOV devices before migration starts, do the migration and then attach similar SRIOV devices on the target.

In general attaching VFIO device (e.g SRIOV VF)  to domain requires extra memory (AFAIK roughly 1GB per device).

When regular VM is created Kubevirt adjusts `libvirtd` rlimits according to the VM spec, when `qemu` process is created
it will inherits its parent rlimits (libvirtd) which are sufficient.

On migrated SRIOV VM's, `qemu` process rlimits are lower then what they should be due to the fact they were detached just before the migration started.
When the migration is finished and SRIOV devices are attached (on tagert VM)  Libvirt adjusts `qemu` process memlock limits accordingly.
Since virt-launcher is non privileged and libvirtd & qemu processes runs with different UID/GID, its necessary to add  `SYS_RESOURCE` to enable Libvirt to adjust `qemu` rlimits.

Libvirt 7 introduce new feature that detects if `qemu` process rlimits are high enough it wont try to change them like it used to, in order to support Kubevirt use case (non-privileged containerized environment).

Now that we use this version its possible to delegate `qemu` process rlimits adjustments to Kubevirt and drop the capability.

ormergi commented:
> Please do not mix rebase with your changes, we have now no way to see the changes done without re-passing through all of them.

I think it it looks like a rebase because I had to change the first commit

ormergi commented:
rebase

AlonaKaplan commented:
> > General question. Why libvirt needs `SYS_RESOURCE` to change the qemu rlimit but kubvirt can do it without this capability?
> 
> virt-handler is priv while virt-launcher is not.
> Before this change, it was already done for the libvirtd process, the only difference here is that we now need the qemu one.

@ormergi can you please add this explanation to the PR message?

EdDev reviewed (approved):
The only open item is the question about the `qemu-system` binary that was raised.
As we do not test it anywhere and I see no ref to it in other places (build components), I think we are fine. But it will be nice to get another view on this.

AlonaKaplan commented:
Thanks Or!
The PR seems fine besides the issue with the qemu process name.
I believe you should have similar code to [here](https://github.com/kubevirt/kubevirt/blob/b2f3a6520ab821dbdf7a7a15f3da24bc32ed449a/cmd/virt-launcher/virt-launcher.go#L562). But let's wait for Vladik's response.

ormergi commented:
> General question. Why libvirt needs `SYS_RESOURCE` to change the qemu rlimit but kubvirt can do it without this capability?
>> virt-handler is priv while virt-launcher is not.
>> Before this change, it was already done for the libvirtd process, the only difference here is that we now need the qemu one.
>
> @ormergi can you please add this explanation to the PR message?

Done

ormergi commented:
There were flakes with some of virt-operator unit tests, the PR to fix it is merged now 
Rebase

vladikr commented on pkg/virt-handler/isolation/detector.go:
```diff
@@ -143,20 +149,82 @@ func (s *socketBasedIsolationDetector) AdjustResources(vm *v1.VirtualMachineInst
+}
+
+// findIsolatedQemuProcess Returns the first occurrence of the QEMU process whose parent is PID"
+func findIsolatedQemuProcess(processes []ps.Process, pid int) ps.Process {
+	processes = childProcesses(processes, pid)
+	return lookupProcessByExecutable(processes, qemuProcessExecutable)
+}
```
I don't see how looking up for the qemu process is different from looking up libvirtd in `AdjustResources`
I'm just curious why do we need this added complexity of this double loop of creating a slice of child processes then looping on it again to fin the executable?
This process looks very simple to me in `AdjustResources`

ormergi replied:
> I don't see how looking up for the qemu process is different from looking up libvirtd in `AdjustResources`
> I'm just curious why do we need this added complexity of this double loop of creating a slice of child processes then looping on it again to fin the executable?
> This process looks very simple to me in `AdjustResources`

@vladikr  Although it look similar finding qemu process is different consider virt-launcher processes state:
```
pid: 2000 ppid: 100 cmd: virt-launcher
|
|_ pid: 3300 ppid: 2000 cmd: qemu-kvm 
|
|_ pid: 4400 ppid: 2000 cmd: virt-launcher   <-------------- isolation detector scope
    |
    |_pid 55000 ppid: 2000 cmd: libvirtd
 ```
- virt-launcher pod runs two virt-launcher processes, where `libvirtd` and `qemu` run's under different instance of virt-launcher process (== both has different parent process).
- Isolation detector, detects the nested virt-launcher process (parent of `libvirtd` process).
Finding `qemu` process requires finding the top-level virt-launcher process, get his child processes and finally look for `qemu-kvm` process.

Finding qemu process requires finding the top-level-virt-launcher pid (the detected virt-launcher process ppid), 
get his child processes and look for qemu-kvm process.

About the complexity,
In my opinion having `qemuProcesses` and `lookupProcessByExecutable` makes the code that looks for qemu process easier to understand (could easily reused for `AdjustResources`).
Doing so made it easier to add tests.

From what I understand and correct me if I am wrong, virt-launcher will run fixed number of processes
thus although having both loops, it shouldn't degrade performance in terms of big-O notation.

What you think?

vladikr replied:
Thank you.

ormergi commented:
Rebase

ormergi commented:
Changes: 
- `findIsolatedQemuProcess` searchs for both qemu-system and qemu-kvm, unit tests has been changed accordingly
- emit warning event when `AdjustQemuProcessMemoryLimits` fails instead of warning log message

vladikr commented:
/retest

vladikr commented:
/approve
Thanks.

vladikr commented:
/lgtm
